File: packages/std/src/math-ops.ts
/**
 * A0 stdlib: math operations
 * math.max, math.min
 */
import type { StdlibFn, A0Record, A0Value } from "@a0/core";

/**
 * math.max { in: list } -> number
 * Returns the maximum value in a numeric list.
 */
export const mathMaxFn: StdlibFn = {
  name: "math.max",
  execute(args: A0Record): A0Value {
    const input = args["in"] ?? null;
    if (!Array.isArray(input)) {
      throw new Error("math.max: 'in' must be a list");
    }
    if (input.length === 0) {
      throw new Error("math.max: list must not be empty");
    }
    let max = -Infinity;
    for (const item of input) {
      if (typeof item !== "number") {
        throw new Error("math.max: all elements must be numbers");
      }
      if (item > max) max = item;
    }
    return max;
  },
};

/**
 * math.min { in: list } -> number
 * Returns the minimum value in a numeric list.
 */
export const mathMinFn: StdlibFn = {
  name: "math.min",
  execute(args: A0Record): A0Value {
    const input = args["in"] ?? null;
    if (!Array.isArray(input)) {
      throw new Error("math.min: 'in' must be a list");
    }
    if (input.length === 0) {
      throw new Error("math.min: list must not be empty");
    }
    let min = Infinity;
    for (const item of input) {
      if (typeof item !== "number") {
        throw new Error("math.min: all elements must be numbers");
      }
      if (item < min) min = item;
    }
    return min;
  },
};

File: go/pkg/stdlib/math_ops.go
package stdlib

import (
	"fmt"
	"math"

	"github.com/thomasrohde/agent0/go/pkg/evaluator"
)

// math.max { in: list } → number
func stdlibMathMax(args *evaluator.A0Record) (evaluator.A0Value, error) {
	input, _ := args.Get("in")
	list, ok := input.(evaluator.A0List)
	if !ok {
		return nil, fmt.Errorf("math.max: 'in' must be a list")
	}
	if len(list.Items) == 0 {
		return nil, fmt.Errorf("math.max: list must not be empty")
	}

	max := math.Inf(-1)
	for _, item := range list.Items {
		num, ok := item.(evaluator.A0Number)
		if !ok {
			return nil, fmt.Errorf("math.max: all elements must be numbers")
		}
		if num.Value > max {
			max = num.Value
		}
	}
	return evaluator.NewNumber(max), nil
}

// math.min { in: list } → number
func stdlibMathMin(args *evaluator.A0Record) (evaluator.A0Value, error) {
	input, _ := args.Get("in")
	list, ok := input.(evaluator.A0List)
	if !ok {
		return nil, fmt.Errorf("math.min: 'in' must be a list")
	}
	if len(list.Items) == 0 {
		return nil, fmt.Errorf("math.min: list must not be empty")
	}

	min := math.Inf(1)
	for _, item := range list.Items {
		num, ok := item.(evaluator.A0Number)
		if !ok {
			return nil, fmt.Errorf("math.min: all elements must be numbers")
		}
		if num.Value < min {
			min = num.Value
		}
	}
	return evaluator.NewNumber(min), nil
}
